enhancements/cluster-logging/content-filter.md

```diff
@@ -0,0 +1,261 @@
+
+### User Stories
+
+* As a logging adminstrator I want to drop log records by severity or other field values in the record.
+* As a logging administrator I want to prune log records by removing pod annotations, or other uninteresting fields.
+
+### Goals
+
+The user can configure the ClusterLogForwarder to drop and prune log records according to their requirements.
+
+### Non-Goals
+
+No complex record transformations, for example adding new fields with computed values.
+
+## Proposal
+
+### Workflow Description
+
+**log administrator** is a human responsible for configuring log collection, storage and forwarding in a cluster.
+
+1. The log administrator creates named `filter` sections in a `ClusterLogForwarder`
+2. Pipelines that require filtering have a `filterRef` field with a list of filter names.
+3. The collector edits forwarded log streams according to the filters.
+
+### API Extensions
+
+A new `filters` section in the `ClusterLogForwarder` allows named filters to be defined.
+This proposal defines two types of fitler "prune" and "drop".
+
+#### Prune filters
+
+Remove all listed fields from each record.
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name:          #user-selected name
+    type: prune
+    prune:
+      remove: []         #remove all fields on this list.
+      removeOthers: []   #remove all fields not on this list.
+```
+
+##### Example
+
+``` yaml
+  spec:
+	filters:
+	- name: foo
+	  type: prune
+	  prune:
+		remove:
+		- kubernetes.flat_labels
+		- pipeline_metadata
+	pipelines:
+	- name: bar
+	  filterRefs: ["foo"]
+```
+
+##### Vector configuration
+
+``` yaml
+[transforms.filter_prune_bar_foo]
+type = "filter"
+inputs = ["*"]
+source =  """
+  del(.kubernetes.flat_labels)
+  del(.pipeline_metadata)
+```
+
+##### Pre-defined filters
+
+Just like default inputs ("application", "audit", "infrastructure") and outputs ("default"), we can define reserved names for pre-defined filters.
+These are equivalent to user-defined filters but pre-defined for convenience.
+
+- "PruneK8sLabels": 
+```yaml
+- name: PruneK8sLabels
+  type: prune
+  prune:
+    remove:
+	  - .kubernetes.flat_labels
+	  - .kubernetes.pod_labels
+	  - .kubernetes.namespace_labels
+	  - .kubernetes.annotations
+
+```
+
+
+#### Drop filters
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name:                #user-selected name
+    type: drop
+    drop:
+	  rules:
+		- path:      # JSON path to the field
+		  action:    # enum: keep|drop
+		  # Following are alternatives, at most one:
+		  in:         # array, true if field value == any element
+		  notIn:      # array, true if field value != all elements
+		  matches:    # regular expression match
+		  notMatches: # regular expression does not match
+		  equals:     # equality - any type
+		  notEquals:  # inequality - any type
+		  less:       # numeric less-than
+		  greater:    # numeric greater-than
+```
+
+Note:
+- Rules are tried in order until a rule matches. Remaining rules are ignored.
```

Comment:
@alanconway  perhaps something like: Rules are tried in order top to down until a rule matches. Remaining rules are ignored? Usually, we understand from top to down, but really, the order can be in both directions

Reply:
Fixed